libyul/optimiser/ReasoningBasedSimplifier.cpp

```diff
@@ -0,0 +1,194 @@
+}
+
+void ReasoningBasedSimplifier::operator()(If& _if)
+{
+	smtutil::Expression condition = encodeExpression(*_if.condition);
+	m_solver->push();
+	m_solver->addAssertion(condition == int2bv(size_t(0)));
+	CheckResult result = m_solver->check({}).first;
+	m_solver->pop();
+	if (result == CheckResult::UNSATISFIABLE)
+		_if.condition = make_unique<yul::Expression>(Literal{locationOf(*_if.condition), LiteralKind::Number, "1"_yulstring, {}});
+
+	m_solver->push();
+	m_solver->addAssertion(condition != int2bv(size_t(0)));
+	CheckResult result2 = m_solver->check({}).first;
+	m_solver->pop();
+	if (result2 == CheckResult::UNSATISFIABLE)
+		// TODO we could actually skip the body is this case
+		_if.condition = make_unique<yul::Expression>(Literal{locationOf(*_if.condition), LiteralKind::Number, "0"_yulstring, {}});
+
+	m_solver->push();
+	m_solver->addAssertion(condition != int2bv(0));
+
+	ASTModifier::operator()(_if.body);
+
+	m_solver->pop();
+}
+
+ReasoningBasedSimplifier::ReasoningBasedSimplifier(
+	Dialect const& _dialect,
+	set<YulString> const& _ssaVariables
+):
+	m_dialect(_dialect),
+	m_ssaVariables(_ssaVariables)
+{
+	m_solver = make_unique<smtutil::Z3Interface>();
+}
+
+smtutil::Expression ReasoningBasedSimplifier::encodeExpression(Expression const& _expression)
+{
+	return std::visit(GenericVisitor{
+		[&](FunctionCall const& _functionCall)
+		{
+			if (auto const* dialect = dynamic_cast<EVMDialect const*>(&m_dialect))
+				if (auto const* builtin = dialect->builtin(_functionCall.functionName.name))
+					if (builtin->instruction)
+						return encodeBuiltin(*builtin->instruction, _functionCall.arguments);
+			return newVariable();
+		},
+		[&](Identifier const& _identifier)
+		{
+			if (
+				m_ssaVariables.count(_identifier.name) &&
+				m_variables.count(_identifier.name)
+			)
+				return m_variables.at(_identifier.name);
+			else
+				return newVariable();
+		},
+		[&](Literal const& _literal)
+		{
+			return int2bv(smtutil::Expression(valueOfLiteral(_literal)));
+		}
+	}, _expression);
+}
+
+smtutil::Expression ReasoningBasedSimplifier::encodeBuiltin(
+	evmasm::Instruction _instruction,
+	vector<Expression> const& _arguments
+)
+{
+	vector<smtutil::Expression> arguments = applyMap(
+		_arguments,
+		[this](Expression const& _expr) { return encodeExpression(_expr); }
+	);
+	switch (_instruction)
+	{
+	case evmasm::Instruction::ADD:
+		return (arguments.at(0) + arguments.at(1));
+	case evmasm::Instruction::MUL:
+		return (arguments.at(0) * arguments.at(1));
+	case evmasm::Instruction::SUB:
+		return (arguments.at(0) - arguments.at(1));
+	case evmasm::Instruction::DIV:
+		return smtutil::Expression::ite(
+			arguments.at(1) == int2bv(0),
+			int2bv(0),
+			arguments.at(0) / arguments.at(1)
+		);
+	// TODO SDIV
+	case evmasm::Instruction::MOD:
+		return smtutil::Expression::ite(
+			arguments.at(1) == int2bv(0),
+			int2bv(0),
+			arguments.at(0) % arguments.at(1)
+		);
+	// TODO SMOD
+	case evmasm::Instruction::LT:
```

Comment:
If SDIV/SMOD is considered, perhaps SLT/SGT should too.

Reply:
Added all of them in the other PR.